### PR TITLE
Implement Azure function rest api getScan

### PR DESCRIPTION
**Pull request by haonliu**
#### Description of changes

Implement Azure function rest api ``getScan``

`getScanBatch` will come in another PR.

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Added relevant unit test for your changes. (`yarn test`)
- [ ] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [ ] Ran precheckin (`yarn precheckin`)

### CODE REVIEW

File: packages/web-api/src/controllers/api-controller.spec.ts [hunk 1]
```diff
@@ -12,13 +14,18 @@ export class ApiControllerMock extends ApiController {
...
         super();
     }
 
     public async handleRequest(): Promise<void> {
         this.handleRequestInvoked = true;
     }
+
+    public getGetRestApiConfigFunc(): () => Promise<RestApiConfig> {
+        return this.getRestApiConfig;
+    }
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/api-controller.spec.ts]: why do you need this?

Comment (devaradhanm) [reply packages/web-api/src/controllers/api-controller.spec.ts]: you could just make this method return super.getRestApiConfig(); & rename this method as getRestApiConfig

File: packages/web-api/src/controllers/scan-result-controller.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        const timeRequested = this.tryGetScanRequestedTime(scanId);
+        if (timeRequested === undefined) {
+            // the scanId is invalid.
+            return;
+        }
+
+        const timeCurrent = new Date();
+        const scanResultQueryBufferInSeconds = (await this.getRestApiConfig()).scanResultQueryBufferInSeconds;
+
+        if (timeCurrent.getTime() - timeRequested.getTime() <= scanResultQueryBufferInSeconds * 1000) {
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/scan-result-controller.ts]: move this calculation to a helper method - isRequestMadeTooSoon

File: packages/web-api/src/controllers/scan-result-controller.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    public async handleRequest(): Promise<void> {
+        const scanId = <string>this.context.bindingData.scanId;
+        const timeRequested = this.tryGetScanRequestedTime(scanId);
+        if (timeRequested === undefined) {
+            // the scanId is invalid.
+            return;
+        }
+
+        const timeCurrent = new Date();
+        const scanResultQueryBufferInSeconds = (await this.getRestApiConfig()).scanResultQueryBufferInSeconds;
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/scan-result-controller.ts]: scanResultQueryBufferInSeconds  => this is not a buffer. rename to minimumWaitTimeforScanStatusCall

File: packages/web-api/src/controllers/api-controller.spec.ts (hunk 1 above)
Comment (haonliu) [reply packages/web-api/src/controllers/api-controller.spec.ts]: updated.

File: packages/web-api/src/controllers/scan-result-controller.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            };
+
+            this.logger.logInfo('scan result fetched', { scanId });
+        }
+    }
+
+    private async getScanResultMapKeyByScanId(scanIds: string[]): Promise<Dictionary<OnDemandPageScanResult>> {
+        const scanResultItems = await this.onDemandPageScanRunResultProvider.readScanRuns(scanIds);
+
+        return keyBy(scanResultItems, item => item.id);
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/scan-result-controller.ts]: you may have to create a not found response for scan id that onDemandPageScanRunResultProvider didnt find

File: packages/web-api/src/controllers/scan-result-controller.ts (hunk 2 above)
Comment (haonliu) [reply packages/web-api/src/controllers/scan-result-controller.ts]: done.

File: packages/web-api/src/controllers/scan-result-controller.ts (hunk 4 above)
Comment (haonliu) [reply packages/web-api/src/controllers/scan-result-controller.ts]: done.

File: packages/web-api/src/controllers/api-controller.spec.ts [hunk 5]
```diff
@@ -3,7 +3,9 @@
 import 'reflect-metadata';
 
 import { Context } from '@azure/functions';
+import { RestApiConfig, ServiceConfiguration } from 'common';
 import { Logger } from 'logger';
+import { Mock, Times } from 'typemoq';
 import { ApiController } from './api-controller';
 
 export class ApiControllerMock extends ApiController {
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/api-controller.spec.ts]: this is not a mock. Call it TestableApiController

File: packages/web-api/src/controllers/api-controller.spec.ts [hunk 6]
```diff
@@ -265,3 +273,30 @@ describe('tryGetPayload()', () => {
         expect(payload).toEqual({ id: 1 });
     });
 });
+
+describe('getRestApiConfig()', () => {
+    it('should call getConfigValue', async () => {
+        const context = <Context>(<unknown>{
+            req: {
+                rawBody: `{ id: 1`,
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/api-controller.spec.ts]: change this value to something else

File: packages/web-api/src/controllers/scan-request-controller.ts [hunk 7]
```diff
@@ -19,7 +20,7 @@ export class ScanRequestController extends ApiController {
         @inject(webApiIocTypes.azureFunctionContext) protected readonly context: Context,
         @inject(ScanDataProvider) private readonly scanDataProvider: ScanDataProvider,
         @inject(GuidGenerator) private readonly guidGenerator: GuidGenerator,
-        @inject(ServiceConfiguration) private readonly serviceConfig: ServiceConfiguration,
+        @inject(ServiceConfiguration) protected readonly serviceConfig: ServiceConfiguration,
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/scan-request-controller.ts]: why did you make this protected?

File: packages/web-api/src/controllers/scan-result-controller.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+            this.logger.logInfo('scan result fetched', { scanId });
+        }
+    }
+
+    private async isRequestMadeTooSoon(scanId: string): Promise<boolean> {
+        const timeRequested = this.tryGetScanRequestedTime(scanId);
+        if (timeRequested === undefined) {
+            // the scanId is invalid.
+            return undefined;
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/scan-result-controller.ts]: should you be returning invalid request  id response or simething instead?

File: packages/web-api/src/controllers/scan-result-controller.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            this.context.res = {
+                status: 422, // Unprocessable Entity,
+                body: `Unprocessable Entity: ${scanId}. ${error}`,
+            };
+        }
+
+        return undefined;
+    }
+
+    private getDefaultResponse(scanId: string): OnDemandPageScanResult {
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/scan-result-controller.ts]: rename this to getTooSoonRequestResponse

File: packages/web-api/src/controllers/scan-result-controller.spec.ts [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+            await scanResultController.handleRequest();
+
+            guidGeneratorMock.verifyAll();
+            expect(context.res.status).toEqual(422);
+            expect(context.res.body).toEqual(`Unprocessable Entity: ${invalidId}. Error: Only version 6 of UUID is supported`);
+        });
+
+        it('should return a default response for requests made too early', async () => {
+            const scanId = 'scanId';
+            context.bindingData.scanId = scanId;
```
Comment (devaradhanm) [inline packages/web-api/src/controllers/scan-result-controller.spec.ts]: refactor test methods by extracting out helper methods. ideally test methods should not be more than ~=10 lines.

File: packages/web-api/src/controllers/scan-request-controller.ts (hunk 7 above)
Comment (haonliu) [reply packages/web-api/src/controllers/scan-request-controller.ts]: This is protected in super class. Same as logger.

File: packages/web-api/src/controllers/scan-result-controller.ts (hunk 8 above)
Comment (haonliu) [reply packages/web-api/src/controllers/scan-result-controller.ts]: It is already been returned in ``tryGetScanRequestedTime``

**Pull request merged**